### ISSUE TITLE
Custom framework Identity object isn't authenticating in core.

### ISSUE THREAD

**Issue by JS-Nati** [dotnet/systemweb-adapters#331]
**Is there a better or more appropriate place for me to ask about this kind of assistance?**

We have custom Identity implementation on our Framework side, and I've been having issues getting the Authorization to work from framework to Core.

It hits the "systemweb-adapters/authenticate" route, and inside that it casts our object to ClaimsPrincipal. And on the Core side all I get is the username, and a ClaimsPrincipal that always has "Context.User.Identity.IsAuthenticated=false". 

Here's our CustomPrincipal object
```c#
public class CustomPrincipal : ICustomPrincipal
{
    public IIdentity Identity { get; private set; }
    public bool IsInRole(string role) 
    {
        return this.Roles.Contains(role, StringComparer.OrdinalIgnoreCase);
    }

    public bool HasRoleIn(params string[] roles)
    {
        foreach(var v in roles)
        {
            if ( IsInRole( v ) ) return true;
        }
        return false;
    }

    public CustomPrincipal(string email)
    {
        this.Identity = new GenericIdentity(email);
    }
    
    //We have a bunch of custom data for our users here as simple properties this is just some that may be relevant
    public int CompanyId { get; set; }
    public int UserId { get; set; }
    public string Username { get; set; }
    public List<string> Roles { get; set; }
    public List<CurrentUserSettings> Settings { get; set; }
}
``` 

I tried modifying the above code and was able to get roles, and userid to come across as claims. I even changed the Identity assignment in the constructor to be '''GenericIdentity(email, "Remote")''' in hopes that it would come across... as IsAuthenticated is a readonly property that uses a simple method to see if X is not null then true.
Alas, nothing I did seems to result in '''Identity.IsAuthenticated=true''' on the core side. 

This is what I get on the Core site routes every time.
"You don't have authorization to view this page.
HTTP ERROR 403"

Please let me know what else you might need to know as I'm not used to posting questions and am trying to present everything i know without being too long winded.

*ASP.NET Framework Application*:
- Technologies and versions used (i.e. MVC): We use a lot, but I know we're using MVC5 because I helped update it to that version from 3.
- .NET Framework Version: 4.7.2
- IIS Version: iisexpress, at the moment... I will have to ask our sysadmin the specifics of our IIS deployment if necessary.
- Windows Version: 10, again... I will have to ask our sysadmin the specifics of our deployment if necessary as it's to a container.

*ASP.NET Core Application*:
- Targeted .NET version: .NET 6
- .NET SDK version: 6

**Comment by JS-Nati** [comment]
I was able to track down what my issue was.

It did have to do with our CustomPrincipal.

I changed the constructor from this:
```c#
    public CustomPrincipal(string email)
    {
        this.Identity = new GenericIdentity(email);
    }
``` 

To this:
```c#
    public CustomPrincipal(string email)
    {
        this.Identity = new ClaimsIdentity(new GenericIdentity(email,AuthenticationTypes.Password));
    }
``` 

This change allowed the IsAuthenticated to come through, and I was able to then view my Core pages.

**Comment by twsouthwick** [comment]
Glad you figured it out.  Since the User object in ASP.NET Core is a `ClaimsPrincipal` and not `IPrincipal`, this should help in migration to stay consistent.

**Comment by tsi-kyle** [comment]
@JS-Nati can you explain how you were able to add the claims from the .NET Framework side?  

We are using the built in roles with SimpleMembership and from .NET 8 I can see that the user is logged in, but roles do not work and the only Claim I have is the username like you mentioned.

I'm thinking I need SystemWebAdapters to convert roles to claims but I'm not sure how to do that.

**Comment by JS-Nati** [comment]
TLDR: All this with the grain of salt. Push comes to shove you need to make sure your User object is able to be cast or converted to a ClaimsPrincipal... the details of accomplishing that in your specific code are understandably out of my purview.



I actually had to download the git repo and step through both sides of it to see where the roles were being lost.
Because we were using a custom principal I basically just needed to make sure that our code generated an object that was compatible with ClaimsPrincipal.

If you want to do something similar, just download/build the code for the version you're currently using, or the main branch, remove the nuget refs from your mvc project, then ref your local build dlls so you can step through and see where it's actually getting lost.

It sounds to me like you may need to do something similar to me and make sure that your user object is a "ClaimsIdentity" or "ClaimsPrincipal" object. It was easy for me because I inherited an existing customprincipal object and just needed to force it to create as the right type of identity object, as my previous code snippets show.
Note: I did not write the custom code we're using. The code I am working with was old and had gone through multiple developers before I started working with it... so my understanding of how Identity works is limited to my troubleshooting on this issue.

If you can't or don't want to pull the repo and deal with stepping through that way, you can throw a breakpoint into one of your controller actions and check your User object in debugger to see what happens when it is cast to "ClaimsPrincipal" object type, because that is what is done when systemweb-adapters sends the User object across to your Core side. The commonality of a ClaimsPrincipal object is how it comes through.

Here's the specific code used in the system-webadapters code where it either casts your User object as ClaimsPrincipal or instantiates a new ClaimsPrincipal passing in your context.User object for conversion in my copy of the main branch, which is admittedly pretty old now:
        // If a user is logged in (using ASP.NET's usual authentication mechanisms), return that claims principal.
        if (context.User.Identity.IsAuthenticated)
        {
            context.Response.StatusCode = 200;
            context.Response.ContentType = "application/octet-stream";
            var claimsPrincipal = context.User as ClaimsPrincipal ?? new ClaimsPrincipal(context.User.Identity);
            using var writer = new BinaryWriter(context.Response.OutputStream);
            claimsPrincipal.WriteTo(writer);
        }

If you throw the highlighted line into one of your controller actions, and make sure you're using your context object, you should be able to see exactly what the adapter is trying to send back to Core, and then you just have to go through your code and see what you need to do to make your User object into something that can be inherited. This update may be done in your Global.asax.cs as that's where you can build application wide interrupts for post authentication code and build custom objects and/or modify your User object, that's where I started tracing back our custom principal object and eventually found where it was being generated and had to figure out what needed to be done to generate a ClaimsPrincipal compatible object. All that being said, I don't know exactly where or how you'll need to modify this, as I was using Custom code and was able to hijack my custom object in its constructor so roles and such were all added after that point.


Hope this helps to at least track things down for you.
--JS-Nati

**Comment by tsi-kyle** [comment]
Wow, thanks for taking the time to explain all of that!

So I did what you said and cast my user as a ClaimsPrincipal and it had no issues and it looks like all of my roles are converted to claims:

![image](https://github.com/dotnet/systemweb-adapters/assets/127792741/76a6021a-c1a3-4bd9-bda3-1c81047f0522)

Doesn't it seem like systemwebadapters should have those claims available on the .NET 8 side?  Maybe I'll have to look into pulling down the package to see how it actually works, like you mentioned.

**Comment by JS-Nati** [comment]
Yeah.

If your current User Object has the appropriate claims when cast as ClaimsPrincipal then you definitely need to trace through things and figure out where the disconnect is.

It’s unfortunate, but that’s just how it works, sometimes. I shared with you the exact code they are using to convert your User to ClaimsPrincipal, so if you can do that successfully  on the framework side, maybe there’s some configuration issue, or something missing on the Core side…

Either way, the easiest way to figure out what is actually happening is to be able to step through the entire process and verify the whole way. Time consuming, but it can be very illuminating.

FYI, Fiddler makes tracking down the appropriate entry points and code on the framework side easier. It’ll show your yarp calls from Core into the Framework auth routes that are created by systemweb-adapters. Again, that’s how I figured out where to put the breakpoints in their code. It’s really only needed to find that, but it’ll get you both the call and the response so you can see exactly what’s sent, break and see how it’s processed, and then what’s coming through in the response.

Sorry if any of this sounds remedial, just a bit of a brain dump on how I figured things out in case it helps someone else… LOL

JS-Nati

**Comment by tsi-kyle** [comment]
Yes you've been a huge help, I really appreciate it.  I'll keep digging in and will report back when I figure out the issue or at the very least a workaround...
